Model registry proposal (#106)
opendatahub-io/opendatahub-community: proposal/model-registry.md

---

> title: Model Registry

Author: I changed the title to ML Model Registry to make the purpose explicit.

---

**Introduction**

> Model Registry in an ML DevOps environment is a central repository for storing metadata about ML models.

Author: Added charter text.

During the development and deployment life cycle of an ML model, various artifacts and metadata are created or associated with the ML model.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer: these are created during training, during serving performance metrics

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer: hyperparameters, metadata

Author: Fixed.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer: an *MLOps Engineer* may work with a versioned Model form repository in serving to provide inferencing on the Model.

Author: Fixed.

Execution of actions and events performed during the model's lifecycle, for example in a pipeline that executes various steps to process and transform data, can also be stored as metadata. 

There needs to be a central repository for all such metadata so that it can provide visibility across the entire lifecycle of the model. 
It should be able to provide lineage and provenance for any artifact or metadata related to an ML model. 

**Design Goals**

Model Registry should be a passive repository which receives and stores ML model related metadata pushed to the registry.
The registry may have simple rules for validating metadata, but it does not take any actions or perform any orchestration based on this metadata. 
The metadata can be pushed to the repository using client libraries in code written by ML users or automatically generated by other ODH components in the platform. For example, metadata about the deployment status of an ML Model. 

> 

Author: Added kubernetes native solution statement.

**User Use Cases**

> - (Optional) *Data Scientists* need self-service access to a Model Registry to store Experiments and Runs for Tracking. 

Reviewer: remove (optional)

*Data Scientists* need access to a Model Registry to store information about Experiments and Runs for Tracking and reproducibility.

Author: Fixed.

- *Data Scientists* can register ML models to share with other *Data Scientists* and *MLOps Engineers*. They can manage versions of registered models and 
- *Data Scientists* and *MLOps Engineers* can design and implement kubeflow pipelines that register their artifacts and executions in the Model Registry. 
- *Data Scientists* and *MLOps Engineers* can browse the Model Registry to deploy ML models to various deployment environments by associating them with deployment stages such as *Stage* and *Production*. 

## Architecture
The high level architecture for an OpenDatahub Model Registry is depicted below.

> ![image](https://github.com/dhirajsb/opendatahub-community/assets/181691/70f45a03-8eef-4e82-98d6-1bf8da1905b6)

Reviewer 2: not sure if this was intentional, but the diagram hyperlink is on another repo? 🤔 we could reference pastebin from PR in case

Author: I edited the file in another repo, so the image must have gotten pasted there. I'll replace it one pasted into the odh repo.

---
title: Model Registry
---

**Introduction**

Model Registry in an ML DevOps environment is a central repository for storing metadata about ML models.

During the development and deployment life cycle of an ML model, various artifacts and metadata are created or associated with the ML model.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Author: Definition from a quick web search: *Hyperparameters* are parameters whose values control the learning process and determine the values of model parameters that a learning algorithm ends up learning. The prefix ‘hyper_’ suggests that they are ‘top-level’ parameters that control the learning process and the model parameters that result from it.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer 3: OK, so it is a general AI term.

I'm not expert, but reading a little bit more on the internet, I find weird the statement telling that "_Model serving will generate metadata about hyper parameters_". Hyper-parameters doesn't seem to be a concern for serving phase, but more for model tuning.

I'm OK with metrics, assuming enhancements are needed because AFAIK, we just expose metrics for consumption.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Author: The Model Registry itself is going to be agnostic as to the semantics or nature and purpose of metadata being stored. 
As long as a shape/schema of metadata is provided to it, it's going to store it in a DB and support retrieving it. So whether they are hyper-parameters or metrics, they are doing to be stored in the repository as some metadata type and it's property values. 
How the metadata is interpreted and used is also up to the client retrieving the metadata.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer 3: OK, I can understand the Model Registry being metadata-agnostic and I'm fine with it.

Yet, the example about _Model Serving generating metadata_ is still misleading to me because, AFAIK, we generating metadata is not in the roadmap for _model serving_.

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Author: I'm with you now. The proposal talks about the general idea that model serving like other ODH components might have metadata that it could expose through the model registry. 
It's not a requirement at the outset, but it could be something it could do in the future. Does that make sense?

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Author: To elaborate, consider the use case where a user downloads a model off of an external repository like huggingface. There is no data scientist training experiment and run to record in the model registry in the beginning. But, is it possible that model serving could introspect the imported model and determine it's hyperparameters and log whats being used for the version being served?

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Reviewer 3: > It's not a requirement at the outset, but it could be something it could do in the future. Does that make sense?

Yes, it does make sense that enhancements on the other components to interact with the model registry may be planned for the future.

> is it possible that model serving could introspect the imported model and determine it's hyperparameters and log whats being used for the version being served?

I do not think this is possible.

---

Re-thinking about this, these are just examples of use-cases. I'm fully :+1: on giving examples, as that helps to realize the value of the model registry.

Perhaps, what is noisy to me is that the hyperparameters example talks about something that cannot be done _at the moment_ and implies that some other enhancements are needed. I'd be more comfortable if the given examples leave aside any enhancement suggestions. If we want to talk about hyperparameters, I'd change the example to something like "the Data Scientist may want to record in the model registry some metadata about the hyperparameters used to train the model", which would be a more realistic example, given current state of our tooling.

So, see this as a nitpick. You don't need to change anything in the proposal. Although if you do, you remove the nit :slightly_smiling_face: .

> For example, a *Data Scientist* may create training data, parameters, metrics during development. During deployment, an *MLOps Engineer* may work with artifacts such as versioned model data. Model serving will generate metadata about hyper parameters and metrics from inference. 

Author: I'll reword it to remove the suggestion that it *has* to be something model serving does. Perhaps the hyperparameters metadata could also come from the model import process while being registered in the model registry.

Execution of actions and events performed during the model's lifecycle, for example in a pipeline that executes various steps to process and transform data, can also be stored as metadata. 

There needs to be a central repository for all such metadata so that it can provide visibility across the entire lifecycle of the model. 
It should be able to provide lineage and provenance for any artifact or metadata related to an ML model. 

**Design Goals**

Model Registry should be a passive repository which receives and stores ML model related metadata pushed to the registry.
The registry may have simple rules for validating metadata, but it does not take any actions or perform any orchestration based on this metadata. 
The metadata can be pushed to the repository using client libraries in code written by ML users or automatically generated by other ODH components in the platform. For example, metadata about the deployment status of an ML Model. 

**User Use Cases**
- (Optional) *Data Scientists* need self-service access to a Model Registry to store Experiments and Runs for Tracking. 

> - *Data Scientists* can register ML models to share with other *Data Scientists* and *MLOps Engineers*. They can manage versions of registered models and 

Reviewer 3: Is the text cut at the end of this line?

Author: Thanks, I was planning on editing the previous sentence, but missed that.

- *Data Scientists* and *MLOps Engineers* can design and implement kubeflow pipelines that register their artifacts and executions in the Model Registry. 

> - *Data Scientists* and *MLOps Engineers* can browse the Model Registry to deploy ML models to various deployment environments by associating them with deployment stages such as *Stage* and *Production*. 

Reviewer 3: For a _use case_ this bullet has some implementation detail that IMHO can be omitted. I would suggest to re-write like:

> *Data Scientists* and *MLOps Engineers* can browse the Model Registry to deploy ML models to various environments (e.g. deployment stages like *Stage* and *Production*, locations like *East*, *West* or specific devices).

(it is just a suggestion, fwiw).

Author: Fixed as suggested.

## Architecture
The high level architecture for an OpenDatahub Model Registry is depicted below.

![image](https://github.com/dhirajsb/opendatahub-community/assets/181691/70f45a03-8eef-4e82-98d6-1bf8da1905b6)

------------------------------------------------------------------------------------------------------------------------------

## Implementation

> WIP

Reviewer 3: Out of curiosity, for other components to interface with the registry, is there any existing standard? or are we going to build something from scratch?

Author: The Model Registry will provide APIs for doing so. The current leading candidate for implementation (pending being blessed from Architecture council) will re-use ml-metadata gRPC API and a implement a higher level GraphQL API.

Pre-trained ML models such as Foundation Models when used directly for model serving will include metadata about their artifacts and hyperparameters. 
Model serving will generate metrics from inference. 
Execution of actions and events performed during the model's lifecycle, for example in a pipeline that executes various steps to process and transform data, can also be stored as metadata. 

There needs to be a central repository for all such metadata so that it can provide visibility across the entire lifecycle of the model. 
It should be able to provide lineage and provenance for any artifact or metadata related to an ML model. 

**Design Goals**

* Model Registry should be a passive repository which receives and stores ML model related metadata pushed to the registry.
* The registry may have simple rules for validating metadata, but it does not take any actions or perform any orchestration based on this metadata. 
* The metadata can be pushed to the repository using client libraries in code written by ML users or automatically generated by other ODH components in the platform. 
For example, metadata about the deployment status of an ML Model. 

> * The Model Registry should also support Kubernetes native deployment just like other Open Data Hub Components. 

Reviewer 4: ```suggestion
* The Model Registry should also support Kubernetes native deployment just like other Open Data Hub Components. 
* It should be also possible to register a model that has not been trained in Open Data Hub
```

**User Use Cases**
- *Data Scientists* need access to a Model Registry to store information about Experiments and Runs for Tracking and reproducibility. 
- *Data Scientists* can register versions of ML models in the Model Registry to share with other *Data Scientists* and *MLOps Engineers*. 
- *Data Scientists* and *MLOps Engineers* can design and implement kubeflow pipelines that register their artifacts and executions in the Model Registry. 
- *Data Scientists* and *MLOps Engineers* can browse the Model Registry to deploy ML models to various environments (e.g. deployment stages like *Stage* and *Production*, locations like *East, West* or specific devices). 

## Architecture

> The high level architecture for an OpenDatahub Model Registry is depicted below.

Reviewer 5: ```suggestion
The high level architecture for an Open Data Hub Model Registry is depicted below.
```

> ![image](https://github.com/opendatahub-io/opendatahub-community/assets/181691/62b9ea30-a1d2-47d6-8f97-c4bfa5ba8a88)

Reviewer 4: Can you replace it with latest version from here https://miro.com/app/board/uXjVMi2jnrI=/ ?

Author: Ramesh has another PR with updates the diagrams that needs to be rebased and merged.

------------------------------------------------------------------------------------------------------------------------------

## Implementation

> WIP

Reviewer 4: What is the plan for this section? Do we expect to have some roadmap/board to link here?

Reviewer 2: ```suggestion
See Trackers on `model-registry` repo:
- https://github.com/opendatahub-io/model-registry/issues/95
```

We could start by adding the first "mvp tracker", wdyt ?

## Owners Information and Maintenance Plan
Ramesh Reddy - RHOAI team - Red Hat

Dhiraj Bokde - RHOAI team - Red Hat

Matteo Mortari - RHOAI team - Red Hat

Andrea Lamparelli - RHOAI team - Red Hat

Isabella do Amaral - RHOAI team - Red Hat

Tony Davidson - RHOAI QE team - Red Hat

The owners are committed to maintaining and testing the Model Registry with each new release of its every component and ODH.

> ## More Info

Reviewer 4: I would link these two tickets
```suggestion
## More Info
- ML-metadata community ticket: https://github.com/google/ml-metadata/issues/185
- Kubeflow community ticket: https://github.com/kubeflow/kubeflow/issues/7396
```
